Add remaining tests of CEReactions for the DOM specification.

rniwa opened the pull request:
Now that we have the basic testing infrastructure for CEReactions, add tests for the remaining interfaces in https://dom.spec.whatwg.org/

Most of tests pass on Chrome, and we intend to fix the behavior in WebKit.

rniwa committed:
Add remaining tests of CEReactions for the DOM specification.

Now that we have the basic testing infrastructure for CEReactions,
add tests for the remaining interfaces in https://dom.spec.whatwg.org/

Most of tests pass on Chrome, and we intend to fix the behavior in WebKit.

dominiccooney commented on custom-elements/reactions/DOMTokenList.html:
```diff
@@ -0,0 +1,223 @@
+<script src="./resources/reactions.js"></script>
+</head>
+<body>
+<div id="log"></div>
```
Is this div necessary?

rniwa replied:
Removed

dominiccooney commented on custom-elements/reactions/DOMTokenList.html:
```diff
@@ -0,0 +1,223 @@
+    assert_equals(logEntries.log().oldValue, null);
+    assert_equals(logEntries.log().newValue, 'foo');
+    assert_equals(logEntries.log().namespace, null);
+}, 'add on DOMTokenList must enqueue a attributeChanged reaction when adding an attribute');
```
Minor grammar point, but maybe "a attributeChanged" &rarr; "an attributeChanged"; below too

rniwa replied:
Oops, fixed that.

dominiccooney commented on custom-elements/reactions/DOMTokenList.html:
```diff
@@ -0,0 +1,223 @@
+    assert_array_equals(element.log().types(), ['constructed']);
+    instance.classList.add('world');
+    assert_array_equals(element.log().types(), []);
+}, 'add on DOMTokenList must enqueue a attributeChanged reaction when adding a value to an existing unobserved attribute');
```
It must *not* enqueue an attributeChanged reaction. The test looks right but the description is confusing.

rniwa replied:
Oops, fixed

dominiccooney commented on custom-elements/reactions/DOMTokenList.html:
```diff
@@ -0,0 +1,223 @@
+    assert_array_equals(element.log().types(), ['constructed']);
+    instance.classList.remove('world');
+    assert_array_equals(element.log().types(), []);
+}, 'remove on DOMTokenList not must enqueue a attributeChanged reaction when removing a value from an unobserved attribute');
```
not must &rarr; must not

There's one other instance of this below, too.

rniwa replied:
Fixed

dominiccooney commented on custom-elements/reactions/DOMTokenList.html:
```diff
@@ -0,0 +1,223 @@
+    assert_equals(logEntries.log().namespace, null);
+}, 'the stringifier of DOMTokenList must enqueue a attributeChanged reaction when the setter is called with the original value of the attribute');
+
+</script>
```
I wonder if we also need tests for eg add('foo', 'bar') that check we only run "update steps" once and hence only one attributeChanged callback?

Another thing to maybe test is an element with a mix of observed and unobserved attributes to ensure the implementation is making the decision about firing attributeChanged on a per-attribute and not per-element basis.

rniwa replied:
Okay, I think we can address that by using non-empty attribute list.

rniwa replied:
And added a test case for adding multiple tokens and removing multiple tokens.

dominiccooney commented on custom-elements/reactions/resources/reactions.js:
```diff
@@ -78,6 +78,38 @@ function testNodeDisconnector(testFunction, name) {
+        assert_array_equals(logEntries.types(), ['constructed', 'attributeChanged']);
+        assert_equals(logEntries.log().name, 'id');
+        assert_equals(logEntries.log().oldValue, null);
+        assert_equals(logEntries.log().newValue, 'foo');
```
Maybe test the namespace parameter here, too?

rniwa replied:
Sure, now that I'm using a helper, that's done.

dominiccooney commented on custom-elements/reactions/DOMTokenList.html:
```diff
@@ -0,0 +1,223 @@
+    instance.classList.add('foo');
+    var logEntries = element.log();
+    assert_array_equals(logEntries.types(), ['attributeChanged']);
+    assert_equals(logEntries.log().name, 'class');
```
This pattern of four assertions is repetitive; I wonder if the tests would be more readable if you add a helper which takes logEntries.log() and four parameters and have the helper do the four assertions?

rniwa replied:
Alright, added assert_attribute_log_entry to reactions.js.

dominiccooney commented on custom-elements/reactions/resources/reactions.js:
```diff
@@ -78,6 +78,38 @@ function testNodeDisconnector(testFunction, name) {
+        assert_array_equals(element.log().types(), ['constructed', 'connected']);
+        var newInstance = testFunction(instance);
+        assert_array_equals(element.log().types(), ['constructed']);
+    }, name + ' must not enqueue a attributeChanged reaction when cloning an element with an unobserved attribute');
```
Is there a way to make this "an element with an unobserved attribute" more accurate but equally succinct? It must not queue an attribute changed callback for *unobserved attributes*; but a custom element must queue an attributeChanged callback for an observed attribute, even if there happen to be other unobserved attributes. Same goes for the earlier file.

rniwa replied:
Sure, I added a new test case. But I think in general, those edge cases are better tested in https://github.com/w3c/web-platform-tests/pull/3853 which targets each reaction type more throughly.  These individual CEReaction tests are just sanity check for each IDL attribute so I don't think we necessary have to exhaustively test each edge case.

dominiccooney reviewed (commented):
Thorough tests! Comments inline.

rniwa committed:
Added assert_attribute_log_entry to reduce the code bloats for validating attribute reactions,
fixed typos, added a test case for adding and removing multiple tokens to DOMTokenList.html,
and added a test case for attributeChanged reactions getting enqueued for only observed attributes
when both unobserved and observed attributes are added during cloning.

dominiccooney reviewed (commented):
LGTM
